File: 7월 18일~/양과 늑대.py
class Node(object):
    def __init__(self, index = None, value = None, left = None, right = None):
        self.index = index
        self.value = value
        self.left = left
        self.right = right

def dfs(graph):
    stack = [(0, 0, 0)]
    answer = 0
    while stack:
        visited = []
        sheep, wolf, position = stack.pop()
        if position in visited:
            continue

        visited.append(position)
        current = graph[position]
        left = current.left
        right = current.right
        
        if current.value == 0:
            sheep += 1
        else:
            wolf += 1
        
        if sheep <= wolf:
            answer = max(answer, sheep)
            continue

        if right and right.value == 1:
            stack.append((sheep, wolf, right.index))
        if left and left.value == 1:
            stack.append((sheep, wolf, left.index))
        if right and right.value == 0:
            stack.append((sheep, wolf, right.index))
        if left and left.value == 0:
            stack.append((sheep, wolf, left.index))
        
    return answer

def solution(info, edges):
    answer = 0
    graph = {}
    for [index, value] in enumerate(info):
        graph[index] = Node(index, value)
    
    for edge in edges:
        A, B = edge
        if graph[A].left is None:
            graph[A].left = graph[B]
        else:
            graph[A].right = graph[B]

    answer = dfs(graph)
    return answer

print(solution([0, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0], [[0, 1], [0, 2], [1, 3], [1, 4], [2, 5], [2, 6], [3, 7], [4, 8], [6, 9], [9, 10]]))

File: 9월 13일/양과늑대.js
function solution(info, edges) {
    let answer = 0;
    let graph = {};
    for (let edge of edges) {
        if (graph.hasOwnProperty(edge[0])) {
            graph[edge[0]].push(edge[1]);
        } else {
            graph[edge[0]] = [edge[1]];
        }
    }

    const DFS = (current, nextNodes) => {
        let [sheep, wolf, currentNode] = current;
        let newNextNodes = [...nextNodes];
        let currentIndex = newNextNodes.indexOf(currentNode);
        info[currentNode] === 0 ? sheep += 1 : wolf += 1;
        answer = Math.max(sheep, answer);
        if (sheep === wolf) return ;

        if (graph.hasOwnProperty(currentNode)) {
            newNextNodes.push(...graph[currentNode]);
        }

        newNextNodes.splice(currentIndex, 1);

        for (let nextNode of newNextNodes) {
            DFS([sheep, wolf, nextNode], newNextNodes);
        }
    };
    DFS([0, 0, 0], [0]);
    return answer;
}

console.log(solution([0, 0, 1, 1, 1, 0, 1, 0, 1, 0, 1, 1], [[0, 1], [1, 2], [1, 4], [0, 8], [8, 7], [9, 10], [9, 11], [4, 3], [6, 5], [4, 6], [8, 9]]))